### ISSUE TITLE
enigmatic "name already defined" error

### ISSUE THREAD

**Issue by xuhdev** [python/mypy#7206]
Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?

Bug.

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

There are four pieces of code:

1.

```python
def func(x: int) -> int:
    return x + 1

def func(x):
    return x
```

2.

```python
def func(x):
    return x

def func(x: int) -> int:
    return x + 1
```

3.

```python
from test import func

def func(x: int) -> int:
    return x + 1
```

4.

```python
def func(x: int) -> int:
    return x + 1

from test import func
```

Whenever test is used as a module, test.py contains

```python
def func(x):
    return x
```

* What is the actual behavior/output?

The first three examples emits "Name 'func' already defined" error, but not the last

* What is the behavior/output you expect?

All four should emit this error.

* What are the versions of mypy and Python you are using?

mypy 0.720+dev.dc6da54f50d5443fe46308046719043eba03c84d  (master at the time of bug report)
Python 3.7.3

* What are the mypy flags you are using? (For example --strict-optional)

None.

**Comment by JukkaL** [comment]
It looks like mypy treats the fourth example as a conditional function definition and generates an error only if the signature of the imported function is incompatible. This is inconsistent with the other cases, so this looks like a bug, though a not very serious one.
